### ISSUE TITLE
Optimise sending data in http3 code

### ISSUE THREAD

**Issue by ddragana** [mozilla/neqo#261]
If a stream is block but still have data we will try to write on every iteration. We should handle this better and wait for stream to be unblocked.

**Comment by hawkinsw** [comment]
I would like to tackle this and wanted to make sure that I have context correct. You are referring to the condition where a stream has data to send but is blocked because it has exhausted its tokens?

```
   If a sender runs out of flow control credit, it will be unable to
   send new data and is considered blocked.  A sender SHOULD send a
   STREAM_DATA_BLOCKED or DATA_BLOCKED frame to indicate it has data to
   write but is blocked by flow control limits.  These frames are
   expected to be sent infrequently in common cases, but they are
   considered useful for debugging and monitoring purposes.
```

from https://tools.ietf.org/html/draft-ietf-quic-transport-22#section-4.1

Thanks for all the great work you're doing on neqo. I hope that I'll be able to help.

**Comment by martinthomson** [comment]
My understanding is that `Http3Transaction::send` is invoked every time the stack is invoked.  That object doesn't track whether the stream is blocked or not, but it could track that and return `false` from `Http3Transaction::has_data_to_send` so that it wouldn't be called every time.

I'm not sure how much of an advantage this is unless we see a lot of blocking.

**Comment by ddragana** [comment]
I think I wanted to change behavior so that a stream is not add to streams_have_data_to_send if it is blocked and only added if it gets unblocked. This will not make a big difference. Also I wanted to change how we consume streams_have_data_to_send. Currently we do
`let to_send = mem::replace(&mut self.streams_have_data_to_send, BTreeSet::new());`
we should actually take single element until it is empty, because a stream can be added again to let self.streams_have_data_to_send during iteration.

**Comment by hawkinsw** [comment]
I completely understand that distinction and that discussion. But, I still wanted to confirm that I am looking in the right part of the spec to understand the context of how a stream would get into the blocked state.

**Comment by martinthomson** [comment]
Yep, you got it right.  A peer provides MAX_STREAM_DATA and MAX_DATA.  Streams are blocked as long as data sent runs to that limit and those values haven't increased.

**Comment by hawkinsw** [comment]
Thanks for confirming. I am really excited about working on this! It's a free-time thing for me, but I will do my best to make progress as quickly as possible!

**Comment by hawkinsw** [comment]
> I think I wanted to change behavior so that a stream is not add to streams_have_data_to_send if it is blocked and only added if it gets unblocked. This will not make a big difference. Also I wanted to change how we consume streams_have_data_to_send. Currently we do
> `let to_send = mem::replace(&mut self.streams_have_data_to_send, BTreeSet::new());`
> we should actually take single element until it is empty, because a stream can be added again to let self.streams_have_data_to_send during iteration.

I have a question! :-)

An http3 server connection's bidi data stream will be marked with `has data to send` when a response is set. While the server is sending that response, it could run out of credits. If/when the streams becomes blocked because it has run out of credits, you are saying that we should remove the stream's `has data to send` marking and only remark it when the stream becomes writable again?

Just wanted to make sure that I properly understand the goal :-)

**Comment by ddragana** [comment]
> > I think I wanted to change behavior so that a stream is not add to streams_have_data_to_send if it is blocked and only added if it gets unblocked. This will not make a big difference. Also I wanted to change how we consume streams_have_data_to_send. Currently we do
> > `let to_send = mem::replace(&mut self.streams_have_data_to_send, BTreeSet::new());`
> > we should actually take single element until it is empty, because a stream can be added again to let self.streams_have_data_to_send during iteration.
> 
> I have a question! :-)
> 
> An http3 server connection's bidi data stream will be marked with `has data to send` when a response is set. While the server is sending that response, it could run out of credits. If/when the streams becomes blocked because it has run out of credits, you are saying that we should remove the stream's `has data to send` marking and only remark it when the stream becomes writable again?
> 
> Just wanted to make sure that I properly understand the goal :-)

yes, that is correct.

**Comment by larseggert** [comment]
@mxinden does your recent work address any of this?

**Comment by mxinden** [comment]
https://github.com/mozilla/neqo/pull/1838 does not. That said, I might tackle it in the future. Thank you for bringing this to my attention.
